crates/glaredb/src/lib.rs

```diff
@@ -0,0 +1,396 @@
+        }
+    }
+
+    fn data_dir(&self) -> Option<PathBuf> {
+        match self.connection_target.clone() {
+            Some(s) => match Url::parse(&s) {
+                Ok(_) => None,
+                Err(_) => Some(PathBuf::from(s)),
+            },
+            None => None,
+        }
+    }
+
+    fn cloud_url(&self) -> Option<Url> {
+        self.connection_target
+            .clone()
+            .and_then(|v| Url::parse(&v).ok())
+    }
+}
+
+#[derive(Clone)]
+pub struct Connection {
+    session: Arc<Mutex<TrackedSession>>,
+    _engine: Arc<Engine>,
+}
+
+impl Connection {
+    pub fn execute(&self, query: impl Into<String>) -> Operation {
+        Operation {
+            op: OperationType::Execute,
+            query: query.into(),
+            conn: Arc::new(self.clone()),
+            schema: None,
+        }
+    }
+
+    pub fn sql(&self, query: impl Into<String>) -> Operation {
+        Operation {
+            op: OperationType::Sql,
+            query: query.into(),
+            conn: Arc::new(self.clone()),
+            schema: None,
+        }
+    }
+
+    pub fn prql(&self, query: impl Into<String>) -> Operation {
+        Operation {
+            op: OperationType::Prql,
+            query: query.into(),
+            conn: Arc::new(self.clone()),
+            schema: None,
+        }
+    }
+}
+
+pub struct RecordStream(Pin<Box<dyn Stream<Item = Result<RecordBatch, DataFusionError>> + Send>>);
+
+impl Stream for RecordStream {
+    type Item = Result<RecordBatch, DataFusionError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.0.poll_next_unpin(cx)
+    }
+}
+
+impl From<SendableRecordBatchStream> for RecordStream {
+    fn from(val: SendableRecordBatchStream) -> Self {
+        RecordStream(val.boxed())
+    }
+}
+
+impl From<Result<SendableRecordBatchStream, DataFusionError>> for RecordStream {
+    fn from(val: Result<SendableRecordBatchStream, DataFusionError>) -> Self {
+        match val {
+            Ok(stream) => stream.into(),
+            Err(err) => RecordStream(Operation::handle_error(err).boxed()),
+        }
+    }
+}
+
+impl From<Result<SendableRecordBatchStream, ExecError>> for RecordStream {
+    fn from(val: Result<SendableRecordBatchStream, ExecError>) -> Self {
+        match val {
+            Ok(stream) => stream.into(),
+            Err(err) => RecordStream(Operation::handle_error(err).boxed()),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+enum OperationType {
+    Sql,
+    Prql,
+    Execute,
+}
+
+#[derive(Clone)]
+#[must_use = "operations do nothing unless call() or execute() run"]
+pub struct Operation {
+    op: OperationType,
+    query: String,
+    conn: Arc<Connection>,
+    schema: Option<Arc<Schema>>,
+}
+
+impl Debug for Operation {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Operation<{:?}>(\"{:?}\")", self.op, self.query)
+    }
+}
```

Comment:
Recommend using the `debug...` methods instead of manually formatting these: https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.debug_struct

Reply:
yeah, I moved this up: because Operation wraps (ultimately)  Engine, the derive-able implementation didn't work. Anyway, I removed it from here and implemented debug for engine in a way that was more reasonable.